From d43a6cbf377faafd06c3f07e977405f654f020c1 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 22 Nov 2012 21:30:59 -0500
Subject: [PATCH] Misc EMC Changes

Only spawn 1 ocelot ever from eggs.
Load custom permissions before plugins
Unload plugins after everything else
Make Durability enchant scale 0-80% reduction on Armor damage (up to level 5)
Zombies break down doors on normal
Update inventory on block place/interact cancellation
Don't update server.properties
Remove spigot update check
Supress empty itemstack log messages
Bat's with Passengers won't sleep
Disable thorns for arrows
Don't update entire inventory for cancelled item damage

as
---
 src/main/java/net/minecraft/server/EntityArrow.java           |  2 +-
 src/main/java/net/minecraft/server/EntityBat.java             |  2 +-
 src/main/java/net/minecraft/server/EntityOcelot.java          |  2 +-
 src/main/java/net/minecraft/server/ItemStack.java             | 11 ++++++++++-
 .../java/net/minecraft/server/PathfinderGoalBreakDoor.java    |  2 +-
 src/main/java/net/minecraft/server/PropertyManager.java       |  2 +-
 .../java/org/bukkit/craftbukkit/event/CraftEventFactory.java  |  2 ++
 7 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 6aa518f..255063c 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -354,7 +354,7 @@ public abstract class EntityArrow extends Entity implements IProjectile {
                 }
 
                 if (entity1 instanceof EntityLiving) {
-                    EnchantmentManager.a(entityliving, entity1);
+                    //EnchantmentManager.a(entityliving, entity1); // EMC
                     EnchantmentManager.b((EntityLiving) entity1, (Entity) entityliving);
                 }
 
diff --git a/src/main/java/net/minecraft/server/EntityBat.java b/src/main/java/net/minecraft/server/EntityBat.java
index 7661544..11b7973 100644
--- a/src/main/java/net/minecraft/server/EntityBat.java
+++ b/src/main/java/net/minecraft/server/EntityBat.java
@@ -133,7 +133,7 @@ public class EntityBat extends EntityAmbient {
             this.yaw += f1;
             if (this.random.nextInt(100) == 0 && this.world.getType(blockposition1).isOccluding()) {
                 // CraftBukkit Start - Call BatToggleSleepEvent
-                if (CraftEventFactory.handleBatToggleSleepEvent(this, false)) {
+                if (!this.isVehicle() && CraftEventFactory.handleBatToggleSleepEvent(this, false)) { // EMC - add !isVehicle
                     this.setAsleep(true);
                 }
                 // CraftBukkit End
diff --git a/src/main/java/net/minecraft/server/EntityOcelot.java b/src/main/java/net/minecraft/server/EntityOcelot.java
index 13c84bd..71d9677 100644
--- a/src/main/java/net/minecraft/server/EntityOcelot.java
+++ b/src/main/java/net/minecraft/server/EntityOcelot.java
@@ -10,7 +10,7 @@ public class EntityOcelot extends EntityTameableAnimal {
     private static final MinecraftKey bI = new MinecraftKey("cat");
     private PathfinderGoalAvoidTarget<EntityHuman> bJ;
     private PathfinderGoalTempt bK;
-    public boolean spawnBonus = true; // Spigot
+    public boolean spawnBonus = false; // Spigot // EMC
 
     public EntityOcelot(World world) {
         super(EntityTypes.OCELOT, world);
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 5373b7c..71274b7 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -402,7 +402,16 @@ public final class ItemStack {
                 j = EnchantmentManager.getEnchantmentLevel(Enchantments.DURABILITY, this);
                 int k = 0;
 
+                j = Math.min(5, j); // EMC
                 for (int l = 0; j > 0 && l < i; ++l) {
+                    // EMC start - make Unbreaking scale to level for armor. 5 = 80% chance of no loss.
+                    if (getItem() instanceof ItemArmor) {
+                        if (random.nextFloat() >= 1 - (.8 * (j / 5))) {
+                            k++;
+                        }
+                        continue;
+                    }
+                    // EMC end
                     if (EnchantmentDurability.a(this, j, random)) {
                         ++k;
                     }
@@ -415,7 +424,7 @@ public final class ItemStack {
                     event.getPlayer().getServer().getPluginManager().callEvent(event);
 
                     if (i != event.getDamage() || event.isCancelled()) {
-                        event.getPlayer().updateInventory();
+                        //event.getPlayer().updateInventory(); // EMC - Handle in Event
                     }
                     if (event.isCancelled()) {
                         return false;
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
index 74da8a9..642f716 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
@@ -43,7 +43,7 @@ public class PathfinderGoalBreakDoor extends PathfinderGoalDoorInteract {
             this.e = i;
         }
 
-        if (this.d == 240 && this.a.world.getDifficulty() == EnumDifficulty.HARD) {
+        if (this.d == 240 && this.a.world.getDifficulty().ordinal() >= EnumDifficulty.NORMAL.ordinal()) { // EMC
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreakDoorEvent(this.a, this.b.getX(), this.b.getY(), this.b.getZ()).isCancelled()) {
                 this.c();
diff --git a/src/main/java/net/minecraft/server/PropertyManager.java b/src/main/java/net/minecraft/server/PropertyManager.java
index 5055a4f..aafdaa3 100644
--- a/src/main/java/net/minecraft/server/PropertyManager.java
+++ b/src/main/java/net/minecraft/server/PropertyManager.java
@@ -72,7 +72,7 @@ public class PropertyManager {
 
         try {
             // CraftBukkit start - Don't attempt writing to file if it's read only
-            if (this.file.exists() && !this.file.canWrite()) {
+            if (this.file.exists()) { // EMC - Never overwrite the file
                 return;
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 7905420..d08ca34 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -198,6 +198,7 @@ public class CraftEventFactory {
         BlockPlaceEvent event = new BlockPlaceEvent(placedBlock, replacedBlockState, blockClicked, item, player, canBuild, equipmentSlot);
         craftServer.getPluginManager().callEvent(event);
 
+        if (event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
@@ -319,6 +320,7 @@ public class CraftEventFactory {
         }
         craftServer.getPluginManager().callEvent(event);
 
+        if (action == Action.RIGHT_CLICK_BLOCK && !cancelledBlock && event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
-- 
2.7.4

